File: Java/340. Longest Substring with At Most K Distinct Characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        
        HashMap<Character, Integer> window = new HashMap<>();
        
        int left = 0;
        int right = 0;
        int dis = 0;
        int count = 0;
        int m = 0;
        
        while(right < s.length()){
            
            char cur = s.charAt(right);
            //System.out.println(cur);    
            if(!window.containsKey(cur) || window.get(cur) == 0){
                window.put(cur, 1);
                dis++;
            }else{
                window.put(cur, window.get(cur)+1);
            }
            count++;
            
            while(dis>k){
                cur = s.charAt(left);
                //System.out.println(cur);
                window.put(cur, window.get(cur)-1);
                count--;
                if(window.get(cur) == 0){
                    dis--;
                }
                left++;
            }
            
            right++;
            m = Math.max(m, count);
            //System.out.println(right);
        }
        
        return m;
    }
}

File: Python3/340. Longest Substring with At Most K Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        window = {}
        
        # left and right index that keep track of the sliding window
        left, right = 0, 0
        
        # distinct chars, total length of the current window, the max length of the windows
        dis, count, m = 0, 0, 0
        
        while right < len(s):
            if s[right] not in window or window[s[right]]==0:
                window[s[right]] = 1
                dis += 1
            else:
                window[s[right]] += 1
            count += 1
            
            while dis>k:
                window[s[left]] -= 1
                count -= 1
                m = max(m, count)
                if window[s[left]] == 0:
                    dis -= 1
                left += 1
            
            right += 1
                
        return max(m, count)
        
        
        